util/dvsim/Scheduler.py

```diff
@@ -233,44 +104,372 @@ def on_sigint(signal_received, frame):
-            # object unnecessarily.
-            tgt_scheduler = self.schedulers.get(item.target)
-            if tgt_scheduler is None:
-                tgt_scheduler = TargetScheduler(item.target)
-                self.schedulers[item.target] = tgt_scheduler
+            add_to_scheduled(item)
 
-            tgt_scheduler.add_item(item)
+        return scheduled
 
-    def run(self):
-        '''Run all items
+    def _remove_from_scheduled(self, item):
+        '''Helper method to remove an item from the _scheduled list.'''
+
+        if item.sim_cfg in self._scheduled[item.target]:
+            if item in self._scheduled[item.target][item.sim_cfg]:
+                self._scheduled[item.target][item.sim_cfg].remove(item)
+
+            if not self._scheduled[item.target][item.sim_cfg]:
+                del self._scheduled[item.target][item.sim_cfg]
+
+    def _get_next_target(self, curr_target):
+        '''Returns the target that succeeds the current one.'''
+
+        if curr_target is None:
+            return next(iter(self._scheduled))
+
+        assert curr_target in self._scheduled
+        target_iterator = iter(self._scheduled)
+        target = next(target_iterator)
+
+        found = False
+        while not found:
+            if target == curr_target:
+                found = True
+            try:
+                target = next(target_iterator)
+            except StopIteration:
+                return None
 
-        Returns a map from item to status.
+        return target
+
+    def _enqueue_successors(self, item=None):
+        '''Move an item's successors from _scheduled to _queued.
+
+        'item' is the recently run job that has completed. If None, then we
+        move all available items in all available cfgs in _scheduled's first
+        target. If 'item' is specified, then we find its successors and move
+        them to _queued.
         '''
-        timer = Timer()
 
-        log.info("[legend]: [Q: queued, D: dispatched, "
-                 "P: passed, F: failed, K: killed, T: total]")
-        results = {}
-        for scheduler in self.schedulers.values():
-            results.update(scheduler.run(timer, results))
-        return results
+        for next_item in self._get_successors(item):
+            assert next_item not in self.item_to_status
+            assert next_item not in self._queued[next_item.target]
+            self.item_to_status[next_item] = 'Q'
+            self._queued[next_item.target].append(next_item)
+            self._remove_from_scheduled(next_item)
+
+    def _cancel_successors(self, item):
+        '''Cancel an item's successors recursively by moving them from
+        _scheduled or _queued to _killed.'''
+
+        items = self._get_successors(item)
+        while items:
+            next_item = items.pop()
+            self._cancel_item(next_item, cancel_successors=False)
+            items.extend(self._get_successors(next_item))
+
+    def _get_successors(self, item=None):
```

Comment:
Not for this PR, but this scheduler seems to be going most of the way to a full dependency DAG. I wonder whether it would be sensible to finish up that transition at some point: I think that some of the code in `_get_successors` and `_get_next_target` would actually get much simpler.

Also, we seem to be re-inventing Make, which is a pity. I wonder whether it would have been more sensible to use Make plus a job dispatch backend. (In a company I used to work for, we could use LSF's "`bmake`", which basically spawned everything through `bsub`. It was really bad manners to use on the queue, because it spawned hundreds of jobs, but it was lightning fast :-))